### ISSUE TITLE
Adjusting `eps` in Chebfun, Chebfun2, and Chebfun3

### ISSUE THREAD

**Issue by trefethen** [chebfun/chebfun#1764]
Traditionally -- i.e. in Chebfun[1] -- we have discouraged users from adjusting `eps` (except when dealing with noisy functions) on the grounds that there is not so much to be gained.  See Section 8.8 of the Guide at http://www.chebfun.org/docs/guide/.

In 2D and even more in 3D, it seems there will be much more to be gained by weakening the tolerance.  Indeed it may happen that we decide to use a default of, say 1e-8, in Chebfun3 -- lots to be thought about here!

Two practical matters come up immediately, more basic than changing the default.  First: both Chebfun2 and Chebfun3 need to accept a specification of `eps` in a call like `f = chebfun2(@(x,y) g(x,y),'eps',1e-8)`.  Action item 1: make this happen in Chebfun2.

Second: both Chebfun2 and Chebfun3 need to make sure that not only each 1D chebfun, but also the overall low-rank compression process respects the current value of Chebfun `eps`.  Currently Chebfun3 (not yet on GitHub... soon we hope!) does this but Chebfun2 does not.  Action item 2: make this happen in Chebfun2.

Here is a first experiment illustrating the modest gain from weakening `eps` in 1D.  The lengths
shorten as `eps` gets looser, and there is some time saved, but it's not a killer.

```
disp('   eps     time    length')
ff = @(x) tanh(100*(x-.3));
for chebeps = 10.^(-16:4:-4)
  chebfunpref.setDefaults('eps',chebeps)
  f = chebfun(ff);
  tic, f = chebfun(ff); t = toc;
  r = length(f);
  fprintf('%8.1e  %6.4f   %4d\n', chebeps,t,r)
end 

   eps     time    length
 1.0e-16  0.0251   2079
 1.0e-12  0.0186   1749
 1.0e-08  0.0188   1099
 1.0e-04  0.0135    377
```

In 2D, at present the improvement in lengths is there but there's no improvement in ranks and timings  because Action item 2 is not yet taken care of.  See the numbers below.  Once Action item 2 is taken care of, both lengths and ranks will diminish as `eps` is loosened and this will add up to a significant savings and speedup.  In Chebfun3 preliminary experiments suggest it will be dramatic indeed.

```
disp('   eps     time      rank m-length n-length')
ff = @(x,y) tanh(10*(x+y-.3));
for chebeps = 10.^(-16:4:-4)
  chebfunpref.setDefaults('eps',chebeps)
  f = chebfun2(ff);
  tic, f = chebfun2(ff); t = toc;
  r = length(f);
  [m,n] = length(f);
  fprintf('%8.1e  %6.4f %7d %7d %8d\n', chebeps,t,r,m,n)
end 

   eps     time      rank m-length n-length
 1.0e-16  1.4684     128     289      257
 1.0e-12  1.4453     128     212      210
 1.0e-08  1.4367     128     146      148
 1.0e-04  1.4389     128      75       75
```

This issue has been written in collaboration with @bhashemi, to whom I will assign it.  Perhaps @ajt60gaibb will be willing to review the fixes to action items 1 and 2 once they are ready?

**Comment by trefethen** [comment]
The above originated from a very particular need: plotting phase portraits, something of interest to @oliviersete as well as myself.  When you have to "smash" a phase portrait as in <http://www.chebfun.org/examples/complex/PortraitsWithPoles.html>, it can be really annoyingly slow.  But if the goal is just a plot to look at, `'eps',1e-5` or some such should be good enough.  Once action item #2 above is taken care of, there's a good chance we'll get a major speedup.

**Comment by nickhale** [comment]
I just looked at implementing this. in principle it should be straight forward (in fact, the only reason it doesn't work right now when passing `eps` through a global preference is related to some intricacy of the preferences structure that I don't understand).

However, in practice, the input parsing of the `chebfun2` constructor is a mess.. which might make this a bit more effort to implement (cleanly).

**Comment by ajt60gaibb** [comment]
In 1D the constructor has complexity  `O(n log n)`, where `n` is the numerical degree of the function.  In 2D the constructor has complexity `O(k^3 + k^2(m+n) + k(m log m + n log n))`, where `m` and `n`are the numerical degrees and `k` is the numerical rank of the function. 

Therefore, there is more to gained in 2D by reducing the numerical rank of a function by increasing `eps`.
Presumably there is even more in 3D.

**Comment by jrus** [comment]
Ideally, making a plot would only bother evaluating enough coefficients and high enough rank to match the required precision for the plot.

In this case, you only care about maybe 3 digits of precision, enough so that the relative magnitude of real and imaginary parts are precise enough to give an accurate color (within the limits of human color perception and 8-bit/channel RGB representation) to their arctangent.

**Comment by bhashemi** [comment]
This can be closed as #1781 is merged a few months ago.
